File: native/cosesign1-validation/src/cose_sign1_verifier.cpp
#include "cosesign1/validation/cose_sign1_verifier.h"

#include <cstdint>
#include <optional>
#include <string>
#include <utility>
#include <vector>

#include "internal/openssl_utils.h"
#include "internal/oqs_utils.h"

#include <cosesign1/common/cbor.h>

namespace cosesign1::validation {

namespace {

using ParsedSign1 = cosesign1::common::cbor::ParsedCoseSign1;

ValidationResult Fail(std::string_view validator_name,
                      std::string message,
                      std::string error_code,
                      std::optional<std::string> property = std::nullopt) {
  ValidationFailure f;
  f.message = std::move(message);
  f.error_code = std::move(error_code);
  f.property_name = std::move(property);
  std::vector<ValidationFailure> failures;
  failures.push_back(std::move(f));
  return ValidationResult::Failure(std::string(validator_name), std::move(failures));
}

std::optional<std::int64_t> ReadAlgFromHeaders(const ParsedSign1& parsed) {
  // COSE rules: protected header parameters are preferred.
  if (const auto prot = parsed.protected_headers.TryGetInt64(1)) {
    return prot;
  }
  return parsed.unprotected_headers.TryGetInt64(1);
}

} // namespace

ValidationResult VerifyCoseSign1(std::string_view validator_name,
                                 const std::vector<std::uint8_t>& cose_sign1,
                                 const VerifyOptions& options) {
  ParsedSign1 parsed;

  std::string parse_error;
  const bool parsed_ok = cosesign1::common::cbor::ParseCoseSign1(cose_sign1, parsed, &parse_error);

  if (!parsed_ok) {
    const std::string msg = parse_error.empty() ? "Invalid COSE_Sign1 structure (CBOR parse failed)"
                                                : ("Invalid COSE_Sign1 structure: " + parse_error);
    return Fail(validator_name, msg, "CBOR_PARSE_ERROR");
  }

  std::optional<std::span<const std::uint8_t>> external_payload;
  std::vector<std::uint8_t> owned_external;
  if (options.external_payload) {
    external_payload = std::span<const std::uint8_t>(options.external_payload->data(), options.external_payload->size());
  } else if (options.external_payload_provider) {
    owned_external = options.external_payload_provider();
    external_payload = std::span<const std::uint8_t>(owned_external.data(), owned_external.size());
  }

  return VerifyParsedCoseSign1(validator_name, parsed, external_payload, options);
}

ValidationResult VerifyParsedCoseSign1(std::string_view validator_name,
                                       const cosesign1::common::cbor::ParsedCoseSign1& parsed,
                                       std::optional<std::span<const std::uint8_t>> external_payload,
                                       const VerifyOptions& options) {

  const auto parsed_alg = ReadAlgFromHeaders(parsed);
  if (!parsed_alg) {
    return Fail(validator_name, "Missing COSE 'alg' header (label 1)", "MISSING_ALG", "alg");
  }

  if (options.expected_alg && static_cast<std::int64_t>(*options.expected_alg) != *parsed_alg) {
    return Fail(validator_name, "COSE 'alg' did not match expected value", "ALG_MISMATCH", "alg");
  }

  std::optional<CoseAlgorithm> alg;
  switch (*parsed_alg) {
    case static_cast<std::int64_t>(CoseAlgorithm::ES256):
      alg = CoseAlgorithm::ES256;
      break;
    case static_cast<std::int64_t>(CoseAlgorithm::ES384):
      alg = CoseAlgorithm::ES384;
      break;
    case static_cast<std::int64_t>(CoseAlgorithm::ES512):
      alg = CoseAlgorithm::ES512;
      break;
    case static_cast<std::int64_t>(CoseAlgorithm::MLDsa44):
      alg = CoseAlgorithm::MLDsa44;
      break;
    case static_cast<std::int64_t>(CoseAlgorithm::MLDsa65):
      alg = CoseAlgorithm::MLDsa65;
      break;
    case static_cast<std::int64_t>(CoseAlgorithm::MLDsa87):
      alg = CoseAlgorithm::MLDsa87;
      break;
    case static_cast<std::int64_t>(CoseAlgorithm::PS256):
      alg = CoseAlgorithm::PS256;
      break;
    case static_cast<std::int64_t>(CoseAlgorithm::RS256):
      alg = CoseAlgorithm::RS256;
      break;
    default:
      return Fail(validator_name, "Unsupported COSE algorithm", "UNSUPPORTED_ALG", "alg");
  }

  std::span<const std::uint8_t> payload_span;
  std::vector<std::uint8_t> owned_external;
  if (parsed.payload) {
    payload_span = std::span<const std::uint8_t>(parsed.payload->data(), parsed.payload->size());
  } else {
    if (!external_payload) {
      if (options.external_payload) {
        external_payload = std::span<const std::uint8_t>(options.external_payload->data(), options.external_payload->size());
      } else if (options.external_payload_provider) {
        owned_external = options.external_payload_provider();
        external_payload = std::span<const std::uint8_t>(owned_external.data(), owned_external.size());
      }
    }
    if (!external_payload) {
      return Fail(validator_name, "Detached payload requires external payload bytes", "MISSING_EXTERNAL_PAYLOAD", "payload");
    }
    payload_span = *external_payload;
  }

  std::optional<std::span<const std::uint8_t>> external_payload_span;
  if (!parsed.payload) {
    external_payload_span = payload_span;
  }

  std::vector<std::uint8_t> tbs;
  (void)cosesign1::common::cbor::EncodeSignature1SigStructure(parsed, external_payload_span, tbs);

  bool ok = false;
  switch (*alg) {
    case CoseAlgorithm::ES256:
    case CoseAlgorithm::ES384:
    case CoseAlgorithm::ES512:
    case CoseAlgorithm::PS256:
    case CoseAlgorithm::RS256: {
      if (!options.public_key_bytes) {
        return Fail(validator_name, "No verification key provided (VerifyOptions.public_key_bytes is required)", "MISSING_KEY");
      }

      auto key = internal::LoadPublicKeyOrCertFromDer(*options.public_key_bytes);
      if (!key) {
        return Fail(validator_name, "Failed to parse public key bytes", "INVALID_PUBLIC_KEY");
      }

      if (*alg == CoseAlgorithm::ES256) {
        ok = internal::VerifyEs256(key.get(), tbs, parsed.signature);
      }
      if (*alg == CoseAlgorithm::ES384) {
        ok = internal::VerifyEs384(key.get(), tbs, parsed.signature);
      }
      if (*alg == CoseAlgorithm::ES512) {
        ok = internal::VerifyEs512(key.get(), tbs, parsed.signature);
      }
      if (*alg == CoseAlgorithm::PS256) {
        ok = internal::VerifyPs256(key.get(), tbs, parsed.signature);
      }
      if (*alg == CoseAlgorithm::RS256) {
        ok = internal::VerifyRs256(key.get(), tbs, parsed.signature);
      }
      break;
    }
    case CoseAlgorithm::MLDsa44:
    case CoseAlgorithm::MLDsa65:
    case CoseAlgorithm::MLDsa87: {
#if defined(COSESIGN1_ENABLE_PQC)
      if (!options.public_key_bytes) {
        return Fail(validator_name, "No PQC verification key provided (set VerifyOptions.public_key_bytes)", "MISSING_KEY");
      }

      ok = internal::VerifyMlDsa(static_cast<std::int64_t>(*alg), tbs, parsed.signature, *options.public_key_bytes);
#else
      (void)tbs;
      return Fail(validator_name, "PQC algorithms are not enabled in this build", "PQC_DISABLED", "alg");
#endif
      break;
    }
  }

  if (!ok) {
    return Fail(validator_name, "Signature verification failed", "SIGNATURE_INVALID");
  }

  std::unordered_map<std::string, std::string> metadata;
  metadata.emplace("alg", std::to_string(*parsed_alg));
  metadata.emplace("payloadLength", std::to_string(payload_span.size()));
  return ValidationResult::Success(std::string(validator_name), std::move(metadata));
}

} // namespace cosesign1::validation

File: rust/cosesign1-validation/src/cose_sign1_verifier.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use cosesign1_common::{encode_signature1_sig_structure, parse_cose_sign1, ParsedCoseSign1};
use ml_dsa::{MlDsa44, MlDsa65, MlDsa87};
use rsa::pkcs1v15;
use rsa::pss;
use rsa::pkcs8::DecodePublicKey as _;
use rsa::RsaPublicKey;
use sha2::Sha256;
use signature::Verifier;
use p256::elliptic_curve::sec1::ToEncodedPoint;
use x509_parser::prelude::FromDer as _;

use crate::validation_result::ValidationResult;

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
#[repr(i64)]
pub enum CoseAlgorithm {
    ES256 = -7,
    ES384 = -35,
    ES512 = -36,
    // Provisional COSE algorithm IDs used by this repo for ML-DSA (post-quantum).
    MLDsa44 = -48,
    MLDsa65 = -49,
    MLDsa87 = -50,
    PS256 = -37,
    RS256 = -257,
}

#[derive(Default, Clone)]
pub struct VerifyOptions {
    pub external_payload: Option<Vec<u8>>,
    pub public_key_bytes: Option<Vec<u8>>,
    pub expected_alg: Option<CoseAlgorithm>,
}

pub fn verify_cose_sign1(validator_name: &str, cose_sign1: &[u8], options: &VerifyOptions) -> ValidationResult {
    let parsed = match parse_cose_sign1(cose_sign1) {
        Ok(p) => p,
        Err(e) => return ValidationResult::failure_message(validator_name, e, Some("COSE_PARSE_ERROR".to_string())),
    };

    let external = options
        .external_payload
        .as_deref()
        .or_else(|| parsed.payload.as_deref());

    verify_parsed_cose_sign1(validator_name, &parsed, external, options)
}

pub fn verify_parsed_cose_sign1(
    validator_name: &str,
    parsed: &ParsedCoseSign1,
    external_payload: Option<&[u8]>,
    options: &VerifyOptions,
) -> ValidationResult {
    let alg = match cose_alg(parsed) {
        Ok(a) => a,
        Err(e) => return ValidationResult::failure_message(validator_name, e, Some("MISSING_OR_INVALID_ALG".to_string())),
    };

    if let Some(expected) = options.expected_alg {
        if expected != alg {
            return ValidationResult::failure_message(
                validator_name,
                format!("alg header mismatch: expected {expected:?}, got {alg:?}"),
                Some("ALG_MISMATCH".to_string()),
            );
        }
    }

    let sig_structure = match encode_signature1_sig_structure(parsed, external_payload) {
        Ok(b) => b,
        Err(e) => return ValidationResult::failure_message(validator_name, e, Some("SIGSTRUCT_ERROR".to_string())),
    };

    let key_bytes = match options.public_key_bytes.as_deref() {
        Some(k) => k,
        None => {
            return ValidationResult::failure_message(
                validator_name,
                "public key bytes not provided",
                Some("MISSING_PUBLIC_KEY".to_string()),
            )
        }
    };

    match verify_signature(alg, key_bytes, &sig_structure, &parsed.signature) {
        Ok(()) => ValidationResult::success(validator_name, Default::default()),
        Err((code, msg)) => ValidationResult::failure_message(validator_name, msg, Some(code)),
    }
}

fn cose_alg(parsed: &ParsedCoseSign1) -> Result<CoseAlgorithm, String> {
    // COSE alg header label is 1.
    let v = parsed
        .protected_headers
        .get_i64(1)
        .or_else(|| parsed.unprotected_headers.get_i64(1))
        .ok_or_else(|| "missing alg header".to_string())?;

    match v {
        -7 => Ok(CoseAlgorithm::ES256),
        -35 => Ok(CoseAlgorithm::ES384),
        -36 => Ok(CoseAlgorithm::ES512),
        -48 => Ok(CoseAlgorithm::MLDsa44),
        -49 => Ok(CoseAlgorithm::MLDsa65),
        -50 => Ok(CoseAlgorithm::MLDsa87),
        -37 => Ok(CoseAlgorithm::PS256),
        -257 => Ok(CoseAlgorithm::RS256),
        _ => Err(format!("unsupported alg: {v}")),
    }
}

fn verify_signature(
    alg: CoseAlgorithm,
    public_key_bytes: &[u8],
    sig_structure: &[u8],
    cose_signature: &[u8],
) -> Result<(), (String, String)> {
    match alg {
        CoseAlgorithm::ES256 => verify_ecdsa_p256(public_key_bytes, sig_structure, cose_signature),
        CoseAlgorithm::ES384 => verify_ecdsa_p384(public_key_bytes, sig_structure, cose_signature),
        CoseAlgorithm::ES512 => verify_ecdsa_p521(public_key_bytes, sig_structure, cose_signature),
        CoseAlgorithm::RS256 => verify_rsa_pkcs1(public_key_bytes, sig_structure, cose_signature),
        CoseAlgorithm::PS256 => verify_rsa_pss(public_key_bytes, sig_structure, cose_signature),
        CoseAlgorithm::MLDsa44 => {
            verify_ml_dsa::<MlDsa44>(public_key_bytes, sig_structure, cose_signature, Some("2.16.840.1.101.3.4.3.17"))
        }
        CoseAlgorithm::MLDsa65 => {
            verify_ml_dsa::<MlDsa65>(public_key_bytes, sig_structure, cose_signature, Some("2.16.840.1.101.3.4.3.18"))
        }
        CoseAlgorithm::MLDsa87 => {
            verify_ml_dsa::<MlDsa87>(public_key_bytes, sig_structure, cose_signature, Some("2.16.840.1.101.3.4.3.19"))
        }
    }
}

fn verify_ml_dsa<P: ml_dsa::MlDsaParams>(
    public_key_bytes: &[u8],
    msg: &[u8],
    sig: &[u8],
    expected_spki_oid: Option<&'static str>,
) -> Result<(), (String, String)> {
    let (encoded_vk_bytes, spki_oid) = extract_ml_dsa_public_key_bytes_and_oid(public_key_bytes)
        .map_err(|e| ("INVALID_PUBLIC_KEY".to_string(), e))?;

    if let (Some(expected), Some(oid)) = (expected_spki_oid, spki_oid.as_deref()) {
        // Only check OID when we were able to parse cert/SPKI.
        if oid != expected {
            return Err((
                "INVALID_PUBLIC_KEY".to_string(),
                format!("unexpected public key algorithm OID: expected {expected}, got {oid}"),
            ));
        }
    }

    let enc_vk = ml_dsa::EncodedVerifyingKey::<P>::try_from(encoded_vk_bytes.as_slice())
        .map_err(|_| ("INVALID_PUBLIC_KEY".to_string(), "bad ML-DSA public key bytes".to_string()))?;
    let vk = ml_dsa::VerifyingKey::<P>::decode(&enc_vk);

    let signature = ml_dsa::Signature::<P>::try_from(sig)
        .map_err(|_| ("BAD_SIGNATURE".to_string(), "bad ML-DSA signature bytes".to_string()))?;

    vk.verify(msg, &signature)
        .map_err(|_| ("BAD_SIGNATURE".to_string(), "signature verification failed".to_string()))
}

fn extract_ml_dsa_public_key_bytes_and_oid(der_or_raw: &[u8]) -> Result<(Vec<u8>, Option<String>), String> {
    // Accept:
    // - DER X.509 certificate
    // - DER SubjectPublicKeyInfo
    // - Raw ML-DSA encoded verifying key bytes
    if let Ok((_, cert)) = x509_parser::parse_x509_certificate(der_or_raw) {
        let spki = &cert.tbs_certificate.subject_pki;
        let oid = Some(spki.algorithm.algorithm.to_string());
        let pk = spki.subject_public_key.data.to_vec();
        return Ok((pk, oid));
    }

    if let Ok((_, spki)) = x509_parser::x509::SubjectPublicKeyInfo::from_der(der_or_raw) {
        let oid = Some(spki.algorithm.algorithm.to_string());
        let pk = spki.subject_public_key.data.to_vec();
        return Ok((pk, oid));
    }

    Ok((der_or_raw.to_vec(), None))
}

fn extract_spki_der_from_der_key_or_cert(der: &[u8]) -> Result<Vec<u8>, (String, String)> {
    if let Ok((_, cert)) = x509_parser::parse_x509_certificate(der) {
        return Ok(cert.tbs_certificate.subject_pki.raw.to_vec());
    }
    Ok(der.to_vec())
}

fn verify_ecdsa_p256(pub_bytes: &[u8], msg: &[u8], sig: &[u8]) -> Result<(), (String, String)> {
    let spki = extract_spki_der_from_der_key_or_cert(pub_bytes)?;
    let pk = p256::PublicKey::from_public_key_der(&spki)
        .map_err(|e| ("INVALID_PUBLIC_KEY".to_string(), format!("bad P-256 public key: {e}")))?;
    let ep = pk.to_encoded_point(false);
    let vk = p256::ecdsa::VerifyingKey::from_sec1_bytes(ep.as_bytes())
        .map_err(|e| ("INVALID_PUBLIC_KEY".to_string(), format!("bad P-256 public key: {e}")))?;
    let signature = p256::ecdsa::Signature::from_slice(sig)
        .map_err(|e| ("BAD_SIGNATURE".to_string(), format!("bad ES256 signature: {e}")))?;
    vk.verify(msg, &signature)
        .map_err(|_| ("BAD_SIGNATURE".to_string(), "signature verification failed".to_string()))
}

fn verify_ecdsa_p384(pub_bytes: &[u8], msg: &[u8], sig: &[u8]) -> Result<(), (String, String)> {
    let spki = extract_spki_der_from_der_key_or_cert(pub_bytes)?;
    let pk = p384::PublicKey::from_public_key_der(&spki)
        .map_err(|e| ("INVALID_PUBLIC_KEY".to_string(), format!("bad P-384 public key: {e}")))?;
    let ep = pk.to_encoded_point(false);
    let vk = p384::ecdsa::VerifyingKey::from_sec1_bytes(ep.as_bytes())
        .map_err(|e| ("INVALID_PUBLIC_KEY".to_string(), format!("bad P-384 public key: {e}")))?;
    let signature = p384::ecdsa::Signature::from_slice(sig)
        .map_err(|e| ("BAD_SIGNATURE".to_string(), format!("bad ES384 signature: {e}")))?;
    vk.verify(msg, &signature)
        .map_err(|_| ("BAD_SIGNATURE".to_string(), "signature verification failed".to_string()))
}

fn verify_ecdsa_p521(pub_bytes: &[u8], msg: &[u8], sig: &[u8]) -> Result<(), (String, String)> {
    let spki = extract_spki_der_from_der_key_or_cert(pub_bytes)?;
    let pk = p521::PublicKey::from_public_key_der(&spki)
        .map_err(|e| ("INVALID_PUBLIC_KEY".to_string(), format!("bad P-521 public key: {e}")))?;
    let ep = pk.to_encoded_point(false);
    let vk = p521::ecdsa::VerifyingKey::from_sec1_bytes(ep.as_bytes())
        .map_err(|e| ("INVALID_PUBLIC_KEY".to_string(), format!("bad P-521 public key: {e}")))?;
    let signature = p521::ecdsa::Signature::from_slice(sig)
        .map_err(|e| ("BAD_SIGNATURE".to_string(), format!("bad ES512 signature: {e}")))?;
    vk.verify(msg, &signature)
        .map_err(|_| ("BAD_SIGNATURE".to_string(), "signature verification failed".to_string()))
}

fn rsa_public_key(pub_bytes: &[u8]) -> Result<RsaPublicKey, (String, String)> {
    let spki = extract_spki_der_from_der_key_or_cert(pub_bytes)?;
    RsaPublicKey::from_public_key_der(&spki)
        .map_err(|e| ("INVALID_PUBLIC_KEY".to_string(), format!("bad RSA public key: {e}")))
}

fn verify_rsa_pkcs1(pub_bytes: &[u8], msg: &[u8], sig: &[u8]) -> Result<(), (String, String)> {
    let key = rsa_public_key(pub_bytes)?;
    let vk = pkcs1v15::VerifyingKey::<Sha256>::new(key);
    let signature = pkcs1v15::Signature::try_from(sig)
        .map_err(|e| ("BAD_SIGNATURE".to_string(), format!("bad RS256 signature bytes: {e}")))?;
    vk.verify(msg, &signature)
        .map_err(|_| ("BAD_SIGNATURE".to_string(), "signature verification failed".to_string()))
}

fn verify_rsa_pss(pub_bytes: &[u8], msg: &[u8], sig: &[u8]) -> Result<(), (String, String)> {
    let key = rsa_public_key(pub_bytes)?;
    let vk = pss::VerifyingKey::<Sha256>::new(key);
    let signature = pss::Signature::try_from(sig)
        .map_err(|e| ("BAD_SIGNATURE".to_string(), format!("bad PS256 signature bytes: {e}")))?;
    vk.verify(msg, &signature)
        .map_err(|_| ("BAD_SIGNATURE".to_string(), "signature verification failed".to_string()))
}
